### PR TITLE
improve cat performance on CPU

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76869
* #76868

This PR is targeting at further improving `torch.cat` performance on CPU.

The original implementation has a fast path for sequential case (1 core), and for parallel case it will:
* if tensors from the input list have the same shape and stride, it will reuse TensorIterator;
* if tensors from the input list list have different shape and stride, it will re-construct TensorIterator for each input tensor.
Both cases have multiple omp sessions which increase parallelization overhead and result in indirect memory access when writing output.

Besides, when the input list is long but each input tensor is small, it will won't go the sequential fast path (result.numel() exceeds grain_size), and each TensorIterator will still go sequentially; so the parallel case will be even slower than the sequential case.

The kernels in this PR is explained in the notes, and briefly:
* when outer_size == 1 (dim=0): parallel on ntensors if the input list is long enough; otherwise parallel on ninputs * input_dim_size;
* when outer_size != 1(dim!=0): when inner_size == 1 or 2, manually vectorize it ; otherwise parallel on outer_size.

All kernels have single omp session, reduce creation of meta info as much as possible.

The following is benchmark result from torch operator_benchmark, Xeon 6248, dual sockets, 20 cores per socket, 2.5GHz.
```bash
numactl --physcpubind=0 --membind=0 python -m pt.cat_test --omp_num_threads 1 --mkl_num_threads 1 --tag_filter ${tag_name}
```
Unit: us per iteration.

### cat_test tag "short"

Name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
cat_sizes(1,1,1)_N2_dim0_cpu | 2.164 | 2.222 | 97.39% | 2.194 | 2.232 | 98.30%
cat_sizes(512,512,2)_N2_dim1_cpu | 315.138 | 315.725 | 99.81% | 91.817 | 88.024 | 104.31%
cat_sizes(128,1024,2)_N2_dim1_cpu | 160.099 | 160.293 | 99.88% | 28.370 | 23.339 | 121.56%

### cat_test tag "long"

Name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
cat_sizes(1024,1024,2)_N2_dim0_cpu | 1599.037 | 1592.922 | 100.38% | 479.180 | 460.884 | 103.97%
cat_sizes(1025,1023,2)_N2_dim1_cpu | 1827.472 | 1653.985 | 110.49% | 703.982 | 470.121 | 149.74%
cat_sizes(1024,1024,2)_N2_dim2_cpu | 7267.629 | 1509.665 | 481.41% | 1259.040 | 444.630 | 283.17%
cat_sizes[<function<lambda>at0x7efe675777a0>,111,65]_N5_dim0_cpu | 1105.257 | 995.440 | 111.03% | 436.243 | 336.996 | 129.45%
cat_sizes[96,<function<lambda>at0x7efe67577680>,64]_N5_dim1_cpu | 812.477 | 789.520 | 102.91% | 239.786 | 218.246 | 109.87%
cat_sizes[128,64,<function<lambda>at0x7efe4af81950>]_N5_dim2_cpu | 1242.302 | 1175.808 | 105.66% | 493.840 | 374.495 | 131.87%
cat_sizes[<function<lambda>at0x7efe4af81b90>,32,64]_N50_dim0_cpu | 2073.221 | 1935.472 | 107.12% | 786.920 | 577.980 | 136.15%
cat_sizes[32,<function<lambda>at0x7efe4af81c20>,64]_N50_dim1_cpu | 2018.251 | 1953.783 | 103.30% | 760.085 | 552.573 | 137.55%
cat_sizes[33,65,<function<lambda>at0x7efe4af81cb0>]_N50_dim2_cpu | 3288.457 | 3135.313 | 104.88% | 1902.809 | 839.656 | 226.62%

(*) the original sequential fast path is good enough except it did not specialize inner_size == 1 or 2.

### cat_test tag "multidim"

name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
cat_sizes(64,32,4,16,32)_N2_dim2_cpu | 4282.034 | 4193.721 | 102.11% | 1251.783 | 1167.688 | 107.20%
cat_sizes(16,32,4,16,32)_N8_dim2_cpu | 4432.973 | 4283.635 | 103.49% | 1225.981 | 1183.442 | 103.59%
cat_sizes(9,31,5,15,33)_N17_dim4_cpu | 6308.275 | 6241.009 | 101.08% | 6514.760 | 1815.949 | 358.75%

### cat_test tag "manyinputs"

name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
cat_sizes[<function<lambda>at0x7f87672d1d40>]_N100_dim0_cpu | 154.795 | 152.816 | 101.30% | 233.893 | 32.100 | 728.64%
cat_sizes[<function<lambda>at0x7f87672d1dd0>]_N1000_dim0_cpu | 256.694 | 252.784 | 101.55% | 849.993 | 148.016 | 574.26%
cat_sizes[<function<lambda>at0x7f87672d1e60>]_N2000_dim0_cpu | 360.112 | 352.116 | 102.27% | 1499.309 | 260.123 | 576.38%
cat_sizes[<function<lambda>at0x7f87672d1ef0>]_N3000_dim0_cpu | 455.611 | 446.080 | 102.14% | 2107.840 | 368.567 | 571.90%

### cat_test tag "static_runtime"

name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
cat_sizes[(1,160),(1,14)]_N-1_dim1_cpu | 2.096 | 2.101 | 99.76% | 2.134 | 2.174 | 98.16%
cat_sizes[(1,20,40),(1,4,40),(1,5,40)]_N-1_dim1_cpu | 2.298 | 2.377 | 96.68% | 2.322 | 2.362 | 98.31%
cat_sizes[(1,580),(1,174)]_N-1_dim1_cpu | 2.161 | 2.241 | 96.43% | 2.186 | 2.219 | 98.51%
cat_sizes[(20,160),(20,14)]_N-1_dim1_cpu | 2.516 | 2.489 | 101.08% | 2.549 | 2.490 | 102.37%
cat_sizes[(20,20,40),(20,4,40),(20,5,40)]_N-1_dim1_cpu | 4.100 | 4.134 | 99.18% | 4.102 | 4.048 | 101.33%
cat_sizes[(20,580),(20,174)]_N-1_dim1_cpu | 3.557 | 3.670 | 96.92% | 3.601 | 3.721 | 96.78%

### transformer shapes

name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
cat_sizes[(5, 8, 113,   64),(5,8,1,64)]_N-1_dim2 | 92.094 | 91.946 | 100.16% | 15.796 | 11.546 | 136.81%

cc @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by mingfeima** [comment]
This PR is targeting at further improving `torch.cat` performance on CPU.

The original implementation has a fast path for sequential case (1 core), and for parallel case it will:
* if tensors from the input list have the same shape and stride, it will reuse TensorIterator;
* if tensors from the input list list have different shape and stride, it will re-construct TensorIterator for each input tensor.
Both cases have multiple omp sessions which increase parallelization overhead and result in indirect memory access when writing output.

Besides, when the input list is long but each input tensor is small, it will won't go the sequential fast path (result.numel() exceeds grain_size), and each TensorIterator will still go sequentially; so the parallel case will be even slower than the sequential case.

The kernels in this PR is explained in the notes, and briefly:
* when outer_size == 1 (dim=0): parallel on ntensors if the input list is long enough; otherwise parallel on ninputs * input_dim_size;
* when outer_size != 1(dim!=0): when inner_size == 1 or 2, manually vectorize it ; otherwise parallel on outer_size.

All kernels have single omp session, reduce creation of meta info as much as possible.

The following is benchmark result from torch operator_benchmark, Xeon 6248, dual sockets, 20 cores per socket, 2.5GHz.
```bash
numactl --physcpubind=0 --membind=0 python -m pt.cat_test --omp_num_threads 1 --mkl_num_threads 1 --tag_filter ${tag_name}
```
Unit: us per iteration.

### cat_test tag "short"

Name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
cat_sizes(1,1,1)_N2_dim0_cpu | 2.164 | 2.222 | 97.39% | 2.194 | 2.232 | 98.30%
cat_sizes(512,512,2)_N2_dim1_cpu | 315.138 | 315.725 | 99.81% | 91.817 | 88.024 | 104.31%
cat_sizes(128,1024,2)_N2_dim1_cpu | 160.099 | 160.293 | 99.88% | 28.370 | 23.339 | 121.56%

### cat_test tag "long"

Name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
cat_sizes(1024,1024,2)_N2_dim0_cpu | 1599.037 | 1592.922 | 100.38% | 479.180 | 460.884 | 103.97%
cat_sizes(1025,1023,2)_N2_dim1_cpu | 1827.472 | 1653.985 | 110.49% | 703.982 | 470.121 | 149.74%
cat_sizes(1024,1024,2)_N2_dim2_cpu | 7267.629 | 1509.665 | 481.41% | 1259.040 | 444.630 | 283.17%
cat_sizes[<function<lambda>at0x7efe675777a0>,111,65]_N5_dim0_cpu | 1105.257 | 995.440 | 111.03% | 436.243 | 336.996 | 129.45%
cat_sizes[96,<function<lambda>at0x7efe67577680>,64]_N5_dim1_cpu | 812.477 | 789.520 | 102.91% | 239.786 | 218.246 | 109.87%
cat_sizes[128,64,<function<lambda>at0x7efe4af81950>]_N5_dim2_cpu | 1242.302 | 1175.808 | 105.66% | 493.840 | 374.495 | 131.87%
cat_sizes[<function<lambda>at0x7efe4af81b90>,32,64]_N50_dim0_cpu | 2073.221 | 1935.472 | 107.12% | 786.920 | 577.980 | 136.15%
cat_sizes[32,<function<lambda>at0x7efe4af81c20>,64]_N50_dim1_cpu | 2018.251 | 1953.783 | 103.30% | 760.085 | 552.573 | 137.55%
cat_sizes[33,65,<function<lambda>at0x7efe4af81cb0>]_N50_dim2_cpu | 3288.457 | 3135.313 | 104.88% | 1902.809 | 839.656 | 226.62%

(*) the original sequential fast path is good enough except it did not specialize inner_size == 1 or 2.

### cat_test tag "multidim"

name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
cat_sizes(64,32,4,16,32)_N2_dim2_cpu | 4282.034 | 4193.721 | 102.11% | 1251.783 | 1167.688 | 107.20%
cat_sizes(16,32,4,16,32)_N8_dim2_cpu | 4432.973 | 4283.635 | 103.49% | 1225.981 | 1183.442 | 103.59%
cat_sizes(9,31,5,15,33)_N17_dim4_cpu | 6308.275 | 6241.009 | 101.08% | 6514.760 | 1815.949 | 358.75%

### cat_test tag "manyinputs"

name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
cat_sizes[<function<lambda>at0x7f87672d1d40>]_N100_dim0_cpu | 154.795 | 152.816 | 101.30% | 233.893 | 32.100 | 728.64%
cat_sizes[<function<lambda>at0x7f87672d1dd0>]_N1000_dim0_cpu | 256.694 | 252.784 | 101.55% | 849.993 | 148.016 | 574.26%
cat_sizes[<function<lambda>at0x7f87672d1e60>]_N2000_dim0_cpu | 360.112 | 352.116 | 102.27% | 1499.309 | 260.123 | 576.38%
cat_sizes[<function<lambda>at0x7f87672d1ef0>]_N3000_dim0_cpu | 455.611 | 446.080 | 102.14% | 2107.840 | 368.567 | 571.90%

### cat_test tag "static_runtime"

name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
cat_sizes[(1,160),(1,14)]_N-1_dim1_cpu | 2.096 | 2.101 | 99.76% | 2.134 | 2.174 | 98.16%
cat_sizes[(1,20,40),(1,4,40),(1,5,40)]_N-1_dim1_cpu | 2.298 | 2.377 | 96.68% | 2.322 | 2.362 | 98.31%
cat_sizes[(1,580),(1,174)]_N-1_dim1_cpu | 2.161 | 2.241 | 96.43% | 2.186 | 2.219 | 98.51%
cat_sizes[(20,160),(20,14)]_N-1_dim1_cpu | 2.516 | 2.489 | 101.08% | 2.549 | 2.490 | 102.37%
cat_sizes[(20,20,40),(20,4,40),(20,5,40)]_N-1_dim1_cpu | 4.100 | 4.134 | 99.18% | 4.102 | 4.048 | 101.33%
cat_sizes[(20,580),(20,174)]_N-1_dim1_cpu | 3.557 | 3.670 | 96.92% | 3.601 | 3.721 | 96.78%


### transformer shapes

name | 1C before | 1C after | 1C speedup | 4C before | 4C after | 4C speedup
-- | -- | -- | -- | -- | -- | --
cat_sizes[(5, 8, 113,   64),(5,8,1,64)]_N-1_dim2 | 92.094 | 91.946 | 100.16% | 15.796 | 11.546 | 136.81%
